### ISSUE TITLE
Modify the RepoPruning pass to delete pruned and no live caller GOs

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-project-prepo#60]
Considering a source file test.c:
```C
static void Ai (void) {}
static void Bi (void) {}
static void Ci (void) { Ai (); }
static void Di (void) { Ai (); Bi (); }
void E (void) { Ci (); }
void F (void) { Di (); }
```
Compile 'test.c' first and then dump all compilations stored in the database. 
```bash
$ rm clang.db
$ clang -O3 -c -target x86_64-pc-linux-gnu-repo test.c -o test.o
$ pstore-dump --all-compilations clang.db
---

- file         :
      path : clang.db
      size : 4194304
  compilations :
      - digest      : ff7792df3b79a9a0ca4f3d34af2fc659
        compilation :
            members :
                - digest     : 094a4b00a820005c4e5996ced264f758
                  fext       : { addr: 328, size: 41 }
                  name       : E
                  linkage    : external
                  visibility : default
                - digest     : d878ff46a2e041dcd8d44f49145ceeb5
                  fext       : { addr: 416, size: 41 }
                  name       : F
                  linkage    : external
                  visibility : default
            path    : \\\\?\\C:\\MyWork\\repo_bug\\TOOLCHAIN-4234
            triple  : x86_64-pc-linux-gnu-repo
...
``` 
Build the test.c a second time without modifying the test.c, and dump the IR code after each opt/llvm passes:
```
$ clang -O3 -c -mllvm -print-after-all -target x86_64-pc-linux-gnu-repo test.c
```
We could see:

1) After repo pruning pass, only E and F will be marked as available externally since only E and F have been stored into the database. Functions Ai, Bi, Ci and Di keep the internal linkage type.
2) All functions go through a lot of passes until reaching the Global Variable Optimizer pass.

We could add an early pruning to remove the available externally functions to avoid unnecessary optimisation work.

In summary:

1) If a function is pruned (already in the database) and has no live ("un-pruned") callers, its IR can be removed (marked as declare) rather than being marked available_externally. This rule covers the functions E and F.  E and F will be removed after repo pruning pass.
2) If function is not pruned (not in the database), has no live ("un-pruned") callers and only visible in defined file (with internal linkage), its IR can be removed (marked as declare) rather than being marked available_externally. This rule covers the functions Ai, Bi, Ci and Di.  Ai, Bi, Ci and Di will be removed.

**Comment by MaggieYingYi** [comment]
The patches haven been committed into the master branch (64ba06dc32e, 0b2a89527d).
